README.md
# Flight Schedules Manager(FSM)

FSM is developed and maintained by the virtual airlines, virtualEgret, and virtual egret is an open source software that others virtual airlines could use to help with their own virtual airlines operations.

## Supported Virtual Airlines Management Software

Below are a list of supported VAM softwares

- phpVMS v7

Other VAMs are not yet planned, but will be considered after phpVMS v7 has been finished.

## Features That are Not Planned

- Subfleet support in flights/pairs
- ACARS functionality

## System Requirements

- Windows 10 or above
- 500mb of RAM
- Needs Network Connection for Functions of Many Features

## Other Requirements

- flightplandatabase.com API key - [Documentation](https://bit.ly/fdbapi)

## Installation

Installation is easy. Download the installer from the releases tab and launch it(no admin rights needed), and the software will be installed

## Features

- Airports
  - Import
    - Import a csv file
      - Currently only support phpVMS exported csv
  - Export
    - Export to a format that phpVMS could read
  - Add
    - Follow a wizard to create an airport
    - Basic airport information is autofilled
  - Clear
    - Remove all imported/stored airports, to start over.
  - Edit
    - Follow a wizard to edit the airports
  - Delete
    - Delete the respective airports
- Flights
  - Pairs Page
    - Innovative 'pairs' system
      - Categorized into departure and arrival ICAO
    - Contains a table of all the departure/arrival pairs
    - Each pair have an option to be edited(flight page)
    - Import
      - Import a phpVMS flight plan
      - Automatically converts into pairs system
    - Export
      - Export all flights to a csv that phpVMS flight page accepts
    - Create New Pair
      - Creates a new pair from a user given departure to a user given arrival
    - Clear
      - Removes all pairs and flights from the application
  - Flight Page
    - List of all flights that starts in the pair departure, and ends in pair arrival

## Support

Support could be found through the discussion option from this repository. More technical support, however, should be raised through the issues feature.

## Bugs

Since this software is maintained and developed by solely one person from the IT team of virtualEgret, bugs are expected. Please raise an issue for bug reports!


changelog.md
# Changelogs

View each version's changelogs here, including dev and beta versions

## v0.2.4-dev1

- Added route searcher function
- Added fail safe to REST requests
- More failsafe will be added soon to other functions and features
- Added change log to keep track of changes
- Added comments to everything


main.js
// * Description: This is the main file of the application, it handles all of the main processes and the main window.

// * Packages Imports
const { app, dialog,BrowserWindow, ipcMain, webContents } = require('electron');
const { autoUpdater } = require('electron-updater');
const prompt = require('electron-prompt');
const isDev = require('electron-is-dev');
const axios = require('axios');
let log = require('electron-log');

// * Default JS Library Imports
var fs = require('fs');
var path = require('path');
const os = require('os');

// ? Checking the appropriate OS the app is ran on
const isMac = os.platform() === "darwin";
const isWindows = os.platform() === "win32";
const isLinux = os.platform() === "linux";


// * Prerun Setup
autoUpdater.allowPrerelease = false;
process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0;
process.env["ELECTRON_DISABLE_SECURITY_WARNINGS"] = "true";

// * Important House Keeping Informations
// 501 = cancelled signals
// 404 = not found
// 401 = unauthorized

// * This is the main window of the application.
let mainWindow;


// * This is the function that creates the main window of the application.
function createWindow () {
  // Logging important information
  log.info("Starting application...")
  log.info(`App version: ${app.getVersion()}`)

  // Setting notifications name to FSM if using windows
  if (isWindows) {
    app.setAppUserModelId("FSM");
  }

  // * Creating the main window
  // ! Changing webPreferences could break the application, be careful.
  mainWindow = new BrowserWindow({
    width: 300,
    height: 500,
    autoHideMenuBar: true,
    icon: __dirname + '/buildResources/icon.png',
    webPreferences: {
      nodeIntegration: true,
      contextIsolation: false,
      enableRemoteModule: true
    },
  });

  // * If under development envronment, load the index.html file, else load the updateCheck.html file.
  if(isDev){
    log.info("Development Environment Detected, skipping update check...")
    mainWindow.setSize(1000, 800);
    mainWindow.loadFile('index.html');
    mainWindow.center();
  }else{
    mainWindow.loadFile('updateCheck.html');
  }

  // * Handling the closing off the application and mainwindow.
  mainWindow.on('closed', function () {
    log.info("Application Closed.")
    mainWindow = null;
  });
}

// * Required pathing data
let downloadPath = app.getPath("downloads");

// * Runs when application is ready.
app.on('ready', () => {
  // Creates the main window
  createWindow();

  // Setting the application to not be resizable.
  log.info("Application Ready, Window no longer resizable.")
  mainWindow.setResizable(false);

  // * Checking for updates only if the application is a stable version.
  log.info("Starting Update Check...")
  if(isStableVersion(app.getVersion())){
    log.info("Production Environment Detected, checking for updates...")
    autoUpdater.checkForUpdatesAndNotify();
  }else{
    log.info("Non-Stable Version Detected, skipping update check...")
    mainWindow.loadFile('index.html');
    mainWindow.setSize(1000, 800);
    mainWindow.center();
  }
});

// * This function checks if the version is a stable version.
function isStableVersion(version){
  version = version.toLowerCase();

  if(version.includes("beta") || version.includes("dev")){
    return false
  }
  return true
}





// * When all windows closes, quit the application for macOS.
app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

// * When app is selected, check if main window exists, if not, create it.
app.on('activate', function () {
  if (mainWindow === null) {
    createWindow();
  }
});





// * This function is called when the application cannot find new updates, therefore it will load the index.html file.
autoUpdater.on('update-not-available', () => {
  log.info("No Updates Available, launching application...")
  mainWindow.setSize(1000, 800);
  mainWindow.loadFile('index.html');
  mainWindow.center();
});

// * This function is called when the application finds new updates, and the update has been downloaded. It will then quit and install the update.
autoUpdater.on('update-downloaded', () => {
  log.info("Update Downloaded, installing application...")
  autoUpdater.quitAndInstall();
});





// * Receiving settings data from the settings file then returns it to the renderer process.
ipcMain.on('getSettings', (event) => {
  log.info("Getting Settings Data...");

  // If settings file exists.
  if(fs.existsSync(path.join(app.getPath("appData"), './FSM/settings.json'))){
    log.info("Settings File Exists, reading...")

    // Checking if the settings file is corrupted.
    let good;
    try{
      JSON.parse(fs.readFileSync(path.join(app.getPath("appData"), './FSM/settings.json'), 'utf8').toString('utf8'))
      good = true
    }catch{
      log.error("Settings File Corrupted, clearing...")
      good = false
    }

    // If the settings file is corrupted, clear it and return a standard JSON. Else, return the settings file's content.
    if(good == false){
      log.info("Settings File Cleared, fixing...")
      fs.writeFileSync(path.join(app.getPath("appData"), './FSM/settings.json'), JSON.stringify({}, null, 2))
      event.returnValue = JSON.stringify({});
    }else{
      log.info("Settings File Read, returning...")
      event.returnValue = fs.readFileSync(path.join(app.getPath("appData"), './FSM/settings.json'), 'utf8').toString('utf8');
    }
  }else{
    log.info("Settings File Does Not Exist, returning just a standard JSON...")

    // If the settings file does not exist, return a standard JSON.
    event.returnValue = JSON.stringify({});
  }
})

// * Clears the settings file, then returning the FS result of the deletion.
ipcMain.on("clearSetup", async (event) => {
  log.info("Clearing Settings Data...");
  event.returnValue = fs.unlinkSync(path.join(app.getPath("appData"), "./FSM/settings.json"))
})

// * Create a dialog with the given parameters.
// @params title: The title of the dialog.
// @params message: The message of the dialog.
ipcMain.on('saveSettings', (event, field, value) => {
  log.info("Saving Settings: " + field + " with value: " + value)

  // If the settings file does not exist, create it.
  if(!fs.existsSync(path.join(app.getPath("appData"), './FSM/settings.json'))){
    fs.writeFileSync(path.join(app.getPath("appData"), './FSM/settings.json'), JSON.stringify({}, null, 2));
  }

  // Read the settings file, parse it, update the field with the value, then write it back to the file. Returns the result of the write.
  let data = fs.readFileSync(path.join(app.getPath("appData"), './FSM/settings.json'), 'utf8').toString('utf8');
  data = JSON.parse(data);
  data[field] = value;
  event.returnValue = fs.writeFileSync(path.join(app.getPath("appData"), './FSM/settings.json'), JSON.stringify(data, null, 2));
})





// * Retrieving flight data from flightplandatabase.com. Returns the data to the renderer process.
// @params from: The ICAO code of the departure airport.
// @params to: The ICAO code of the arrival airport.
ipcMain.on("findFlightData", async (event, from, to) => {
  var config = {
    method: 'get',
    url: 'https://api.flightplandatabase.com/search/plans?toICAO=' + to + '&fromICAO=' + from,
  };
  
  // Error handler, provides a fail safe.
  let planSearch
  try{
    planSearch = await axios(config)
  }catch(e){
    planSearch = {
      data: e.response.status
    }
  }

  event.returnValue = planSearch;
  
})

// * Creating flight data from flightplandatabase.com. Returns the data to the renderer process.
// @params from: The ICAO code of the departure airport.
// @params to: The ICAO code of the arrival airport.
ipcMain.on("createFlightData", async (event, from, to) => {
  // Inside so the settings is updated every time the function is called.
  let settings = require(path.join(app.getPath("appData"), './FSM/settings.json'))


  var data = JSON.stringify({
    "fromICAO": from,
    "toICAO": to
  });
  
  var config = {
    method: 'post',
    url: 'https://api.flightplandatabase.com/auto/generate',
    headers: { 
      'Authorization': 'Basic ' + settings["fpdAPI"], 
      'Content-Type': 'application/json'
    },
    data : data
  };

  // Error handler, provides a fail safe.
  let planSearch
  try{
    planSearch = await axios(config)
  }catch(e){
    planSearch = {
      data: e.response.status
    }
  }
  planSearch = planSearch.data;

  event.returnValue = planSearch;
})

// * Retrieving flight route from flightplandatabase.com. Returns the data to the renderer process.
// @params id: The ID of the flightplandatabase.com flight plan.
ipcMain.on("findFlightRoute", async (event, id) => {
  var config = {
    method: 'get',
    url: 'https://api.flightplandatabase.com/plan/' + id,
  };

  // Error handler, provides a fail safe.
  let route
  try{
    route = await axios(config)
  }catch(e){
    route = {
      data: e.response.status
    }
  }
  route = route.data;

  event.returnValue = route["route"]["nodes"];
  
})





// * Create a prompt dialog with the given parameters.
// @params title: The title of the prompt dialog.
// @params label: The label of the prompt dialog.
// @params value: The value of the prompt dialog.
// @params type: The type of the prompt dialog.
ipcMain.on('dialogCreate', async (event, title, label, value, type) => {
  log.info("Prompt Requested: " + title + " with label: " + label + " and value: " + value + " and type: " + type)

  // Using the prompt addon and the parameters to create the dialog
  let data = await prompt({
      title: title,
      label: label,
      type: type,
      value: value,
      alwaysOnTop: true,
      icon: __dirname + '/buildResources/icon.png',
    })

    // Result handler, if the user cancels the dialog, return 501, if user did not fill prompt in, return 404, else return the data.
    let res;
    if(data == null){
      res = 501;
    }
    if(res != 501 && data == ""){
      res = 404;
    }
    if(res != 501 && res != 404)res = data;
    event.returnValue = res;
});

// * This function generates a dialog popup for the user.
// Returns: the user's input.
ipcMain.on('dialog', async (event, method, params) => {  
  log.info("Dialog Requested: " + method + " with params: " + JSON.stringify(params))    
  let data = await dialog[method](params);
  event.returnValue = data;
});

// * Create a REST Request.
// @params url: The URL of the REST Request.
// @params method: The method of the REST Request.
// @params header: The header of the REST Request.
ipcMain.on("RESTreq", async (event, url, method, header) => {
  log.info("REST Requested: " + url + " with method: " + method + " and header: " + header)

  var config = {
    method: method,
    url: url,
    headers: header
  };

  // Error handler, provides a fail safe.
  try{
    let result = await axios(config);
    event.returnValue = result.data
  }catch(e){
    event.returnValue = e.response.status;
  }
  
})





// * Get information for an airport from aviowiki.com.
// @params icao: The ICAO code of the airport.
ipcMain.on('getAirportInfo', async (event, icao) => {
  log.info("Getting Airport Info for: " + icao)
  var config = {
    method: 'get',
    maxBodyLength: Infinity,
    url: 'http://api.aviowiki.com/free/airports/search?query=' + icao,
    headers: { }
  };
  
  // Error handler, provides a fail safe.
  let data
  try{
    data = await axios(config)
  }catch(e){
    data = {
      data: e.response.status
    }
  }

  log.info(icao + " response: " + JSON.stringify(data.data))

  // If the airport does not exist, return 404.
  if(data.data.content[0] == undefined || data.data.content[0].icao == null){
    event.returnValue = 404
    return
  }
  
  // If the airport exists, return the data.
  event.returnValue = data.data.content[0]
}); 


// * Create a dialog with the given parameters. Used for adding/editing airports.
// @params rowNum: The row number for the airport.
// @params name: The field for the edit.
// @params value: The value of the edit.
// @params require: If the field is required.
// @params defaults: The default value of the field.
ipcMain.on('editAirportDialog', async (event, rowNum, name, value, require, defaults) => {
  log.info("Editing Airport: " + name + " with value: " + value + " and require: " + require + " and defaults: " + defaults)
  let data = await prompt({
    title: 'Edit Airport',
    label: 'Enter the value for \'' + name + '\'',
    value: value,
    type: 'input',
    icon: __dirname + '/buildResources/icon.png',
  })

  // Result handler, if the user cancels the dialog, return 501, if user did not fill prompt in, return the default value(if field is required), else return the data.
  let done = false;
  if(done == false && data == null){
    event.returnValue = { data: 501, rowNum: rowNum };
    done = true
  }
  if(done == false && require == true && data == ""){
    event.returnValue = { data: defaults, rowNum: rowNum };
    done = true
  }
  if(data == undefined)data = "";

  if(done == false){
    event.returnValue = { data: data, rowNum: rowNum };
  }
});

// * Function to read the imported CSV Airport file.
// @params file: The file to read.
ipcMain.on('readAirportImport', (event, file) => {
  log.info("Reading Airport Import: " + file)
  var data = fs.readFileSync(path.join(file[0]), 'utf8');

  // Processing the data so renderer can read it.
  data = data.split("\n")
  
  for(i in data){
    data[i] = data[i].split(",")
  }
  log.info("CSV Data: " + data)
  event.sender.send('readAirportImport', { data: data });
});

// * Function to export the airport data to a CSV file.
// @params file: The file to export to.
ipcMain.on('exportAirport', (event, file) => {
  log.info("Exporting Airport Data: " + file)
  file = JSON.parse(file);
  for(i in file){
    file[i] = file[i].join(",")
  }
  file = file.join("\n")

  
  log.info("Converted CSV Data: " + file)

  fs.writeFile(path.join(downloadPath, "./exportedAirport.csv"), file, 'utf8', function (err) {
    if (err) {
      log.error(err)
      event.sender.send('exportAirport', { error: true });
    } else{
      event.sender.send('exportAirport', { error: false });
    }
  });
});





// * Function to read the imported CSV Flights file.
// @params file: The file to read.
ipcMain.on('readPairImport', (event, file) => {
  log.info("Reading Pairs Import: " + file)
  var data = fs.readFileSync(path.join(file[0]), 'utf8');
  data = data.split("\n")
  
  for(i in data){
    data[i] = data[i].split(",")
  }
  log.info("CSV Data: " + data)
  event.sender.send('readPairImport', { data: data });
});

// * Function to export the flight data to a CSV file.
// @params file: The file to export to.
ipcMain.on('exportFlights', (event, file) => {
  file = JSON.parse(file);
  for(i in file){
    file[i] = file[i].join(",")
  }
  file = file.join("\n")
  log.info("Exporting CSV Data: " + file)

  fs.writeFile(path.join(downloadPath, "./exportedFlights.csv"), file, 'utf8', function (err) {
    if (err) {
      log.error(err)
      event.sender.send('exportFlights', { error: true });
    } else{
      event.sender.send('exportFlights', { error: false });
    }
  });
});

// * Creates a prompt used for editing/adding pairs.
// @params rowNum: The row number for the pair.
// @params name: The field for the edit.
// @params value: The value of the edit.
// @params require: If the field is required.
// @params defaults: The default value of the field.
ipcMain.on('editPairs', async (event, rowNum, name, value, require, defaults) => {
  log.info("Editing Pairs: " + name + " with value: " + value + " and require: " + require + " and defaults: " + defaults)
  let data = await prompt({
      title: 'Edit Pairs',
      label: 'Enter the value for \'' + name + '\'',
      value: value,
      type: 'input',
      icon: __dirname + '/buildResources/icon.png',
    })

    // Result handler, if the user cancels the dialog, return 501, if user did not fill prompt in, return the default value(if field is required), else return the data.
    let done = false;
    if(done == false && data == null){
      event.returnValue = { data: 501, rowNum: rowNum };
      done = true
    }
    if(done == false && require == true && data == ""){
      event.returnValue = { data: defaults, rowNum: rowNum };
      done = true
    }
    if(data == undefined)data = "";
  if(done == false){
    event.returnValue = { data: data, rowNum: rowNum };
  }
});


pages/js/pairs.js
// * Description: The javascript for the pairs page


// * Load the flight data from the local storage
$(function() {
  let data = JSON.parse(window.localStorage.getItem("airports"));
  if(data.length <= 1){
    alert("Please add an airport first!");
    window.location.href = "airports.html";
  }
  load(JSON.parse(window.localStorage.getItem("pairs")));
})


// * Export the flights
// @param data - The data from the import
ipcRenderer.on("exportFlights", (event, data) => {
  if(data.error == true){
    alert("An error occured!")
  }else{
    alert("Exported!")
  }
})

// * Read pairs imports
// @param data - The data from the import
ipcRenderer.on("readPairImport", (event, data) => {
  data = data.data;
  window.localStorage.setItem("flights", JSON.stringify(data));
  data.shift();

  // Get the pairs and get the unique pairs
  let temps = ['dep_icao,arr_icao,flights'.split(",")];
  for(i in data){
      if(data[i][5] == undefined || data[i][6] == undefined)continue;
      if(checkExist(temps, data[i][5], data[i][6]) == false){
        temps.push([data[i][5], data[i][6], []])
      }
  }

  // Add the flights to the pairs
  for(i in data){
      if(data[i][5] == undefined || data[i][6] == undefined)continue;
      for(j in temps){
          if(temps[j][0] == data[i][5] && temps[j][1] == data[i][6]){
              temps[j][2].push(data[i][0] + data[i][1])
          }
      }
  }

  // Save the pairs to the local storage
  window.localStorage.setItem("pairs", JSON.stringify(temps));
  load(JSON.parse(window.localStorage.getItem("pairs")));
})

// * Function to check if the pair already exists
async function importPair(){
  // Check if the user wants to continue without saving
  let confirmed = false;
  if(document.getElementById("pairsInformation").rows.length > 1){
      let temp = await confirm("Are you sure you want to continue without saving?")
      confirmed = temp == 0 ? true : false;
  }else{
      confirmed = true;
  }
  if(confirmed == false)return

  // Open the file dialog
  let files = ipcRenderer.sendSync("dialog", "showOpenDialog", {
    properties: ['openFile'],
    filters: [
        { name: 'phpVMS Airports Export CSV', extensions: ['csv'] }
    ]
  })

  // Check if the user selected a file
  if (files !== undefined) {
      // handle files
      ipcRenderer.send('readPairImport', files.filePaths)
  }
}

// * Function to exports the pairs
function exportFlights(){
  // Check if there is anything to export
  var oRows = document.getElementById('pairsInformation').getElementsByTagName('tr');
  var iRowCount = oRows.length;
  if(iRowCount == 1){
      alert("There is nothing to export!")
      return;
  }

  // Export the flights
  let information = window.localStorage.getItem("flights");
  ipcRenderer.send('exportFlights', information)
}

// * Function to load the pairs
function clearPairs(){
  // Check if the user wants to clear the pairs
  let warn = confirm("Are you sure you want to clear all pairs? This cannot be undone!");
  if(warn == 1)return;

  // Clear the pairs
  window.localStorage.setItem("flights", JSON.stringify(['airline,flight_number,route_code,callsign,route_leg,dpt_airport,arr_airport,alt_airport,days,dpt_time,arr_time,level,distance,flight_time,flight_type,load_factor,load_factor_variance,pilot_pay,route,notes,start_date,end_date,active,subfleets,fares,fields'.split(",")]));
  window.localStorage.setItem("pairs", JSON.stringify(['dep_icao,arr_icao,flights'.split(",")]));

  // Load the pairs
  load(JSON.parse(window.localStorage.getItem("pairs")));
}

// * Function to load the pairs
function createPair(){
  // Get the departure and arrival airports
  let dep = ipcRenderer.sendSync('editPairs', 0, "DepartureICAO(rqd)", "", true, "");
  dep = dep.data;
  if(dep == 501)return;
  if(dep == "")return alert("Departure Airport cannot be empty!");
  let arr = ipcRenderer.sendSync('editPairs', 0, "ArrivalICAO(rqd)", "", true, "");
  arr = arr.data;
  if(arr == 501)return;
  if(arr == "")return alert("Departure Airport cannot be empty!");

  // Check if the airports exist and if the pair already exists
  if(airportExist(dep) == false)return alert("Departure Airport not found!");	
  if(airportExist(arr) == false)return alert("Arrival Airport not found!");
  if(pairExist(dep, arr) == true)return alert("Pair Already Exists!");

  // Add the pair
  let warn = confirm("Are you sure you want to add this pair? Added pairs cannot be individually deleted!");
  if(warn == 1)return;
  let data = JSON.parse(window.localStorage.getItem("pairs"));
  data.push([dep.toUpperCase(), arr.toUpperCase(), []]);
  window.localStorage.setItem("pairs", JSON.stringify(data));

  // Load the pairs
  load(data);
}

// * Function to load the pairs
// @params data - The data to load
// @params dep - The ICAO of the departure airport
// @params arr - The ICAO of the arrival airport
function checkExist(data, dep, arr){
  let res = false
  for(i in data){
    if(dep == undefined || arr == undefined)continue;
      if(data[i][0].toString().toLowerCase() == dep.toString().toLowerCase() && data[i][1].toString().toLowerCase() == arr.toString().toLowerCase()){
          res = true;
          break;
      }
  }
  return res;
}

// * Function to check if the pair already exists
// @params dep - The ICAO of the departure airport
// @params arr - The ICAO of the arrival airport
function pairExist(dep, arr){
  let data = JSON.parse(window.localStorage.getItem("pairs"));
  let res = false
  for(i in data){
      if(data[i][0].toString().toLowerCase() == dep.toString().toLowerCase() && data[i][1].toString().toLowerCase() == arr.toString().toLowerCase()){
          res = true;
          break;
      }
  }
  return res;
}

// * Function to check if the airport exists
// @params icao - The ICAO of the airport
function airportExist(icao){
  let data = JSON.parse(window.localStorage.getItem("airports"));
  let res = false
  for(i in data){
      if(data[i][0].toString().toLowerCase() == icao.toString().toLowerCase()){
          res = true;
          break;
      }
  }
  return res;
}

// * Function to load the pairs
// @params data - The data to load
function load(data) {
  // Reset the table
  document.getElementById("pairsInformation").innerHTML = `
<tr>
  <th>Depature ICAO</th>
  <th>Arrival ICAO</th>
  <th># of Included Flights</th>
  <th>Actions</th>
</tr>
    `

  // Load the pairs into the table
  for(i in data){
        if(i != 0 && data[i].length > 1){
            document.getElementById("pairsInformation").innerHTML += `
<tr>
  <td>${data[i][0]}</td>
  <td>${data[i][1]}</td>
  <td>${data[i][2].length.toString()}</td>
  <td><button onClick="editPair(${i})">Edit Pair</button></td>
</tr>
            `
        }
    }
}


// * Function to edit the pair
// @params id - The id of the pair
function editPair(id){
  let data = JSON.parse(window.localStorage.getItem("pairs"));
  window.localStorage.removeItem("editPair");
  window.localStorage.setItem("editPair", JSON.stringify(data[id]));
  window.location.href = './editPairs.html'
}


pages/js/editPairs.js
// * Description: The javascript file for the edit pairs page


// * Load the flight data from the local storage
$(function() {
  let data = JSON.parse(window.localStorage.getItem("editPair"));
  if(data.length <= 1){
    window.location.href = "./pairs.html";
  }
  load(JSON.parse(window.localStorage.getItem("editPair")));
})

// * Function to go back to the pairs page
function back(){
    window.localStorage.removeItem("editPair")
    window.location.href = "./pairs.html"
}

// * Function to edit the flight
// @params index - The index of the flight in the local storage
function editFlight(index){
    let flightRoute = getFlightRoute(index);
    if(typeof flightRoute == "number")return alert("An error(" + flightRoute + ") occured while trying to get the flight route. Please try again later.");
    console.log(flightRoute)
}

// * Function to load the flight data
// @param data - The data from the local storage
function load(data){
    // Get the flights from the local storage
    let flights = JSON.parse(window.localStorage.getItem("flights"));

    // Get the flights information table
    document.getElementById("flightsInformation").innerHTML = `
    <tr>
        <th>Flight Number</th>
        <th>Code</th>
        <th>Leg</th>
        <th>Departure ICAO</th>
        <th>Arrival ICAO</th>
        <th>Actions</th>
    </tr>
    `

    // Loop through the flights and add them to the table
    let fltNum = data[2];
    for(j in fltNum){
        for(i in flights){
            if((flights[i][0]+flights[i][1]).toString().toLowerCase() == fltNum[0].toString().toLowerCase()){
                document.getElementById("flightsInformation").innerHTML += `
                <tr>
                    <td>${flights[i][0]+flights[i][1]}</td>
                    <td>${flights[i][2]}</td>
                    <td>${flights[i][4]}</td>
                    <td>${flights[i][5]}</td>
                    <td>${flights[i][6]}</td>
                    <td><button onClick="editFlight(${i})">Edit Flight</button></td>
                </tr>
                `
            }
        }
    }
}


// * Function to get the flight route
// @params index - The index of the flight in the local storage
function getFlightRoute(index){
    // Get the flights from the local storage and find the flight data
    let flights = JSON.parse(window.localStorage.getItem("flights"));
    let flightData = ipcRenderer.sendSync("findFlightData", flights[index][5], flights[index][6]);
    // check if there is an error
    if(typeof flightData == "number")return flightData;

    // Get the flight id, while trying to safe more rate limits for the user.
    let flightId;
    if(flightData.length == 0){
        let newFlightData = ipcRenderer.sendSync("createFlightData", flights[index][5], flights[index][6]);
        if(typeof newFlightData == "number")return newFlightData;
        flightId = newFlightData["id"];
    }else{  
        // Get the most popular flight route
        let top = 0;
        let topIndex = 0;
        for(i in flightData){
            let data = flightData[i];
            if(data.popularity > top){
                top = data.popularity;
                topIndex = i;
            }
        }

        // Get the flight id
        flightId = flightData[topIndex]["id"];
    }

    // Get the flight route    
    let routeObject = ipcRenderer.sendSync("findFlightRoute", flightId)
    if(typeof routeObject == "number")return routeObject;

    // Convert the route object to a string
    let route = "";
    for(i in routeObject){
        route += routeObject[i]["ident"] + " "
    }

    // Remove the last space
    route = route.substring(0, route.length-1)

    // Return the flight route
    return route;
}


pages/js/default.js
// * Description: This file is the main file for the application. It is the file that is loaded when the application is launched.


// * Import all neccessary modules
const { ipcRenderer, dialog } = require('electron');
let $ = jQuery = require('jquery');
var axios = require('axios');
const { type } = require('jquery');

// * Setting required localStorages, and checking if settings satisfy requirements
$(async function() {
  // Added notices to dev console
  console.log("DOING ANYTHING HERE WILL PROBABLY BREAK THE APPLICATION!")
  window.localStorage.setItem("NOTICE", "EDITS MADE TO ANY OF THE FOLLOWING FIELDS WILL BREAK THE APPLICATION!");

  // If airport localStorage is empty, then set it to the default value
  if(window.localStorage.getItem("airports") == null){
    window.localStorage.setItem("airports", JSON.stringify(['icao,iata,name,location,country,timezone,hub,lat,lon,ground_handling_cost,fuel_100ll_cost,fuel_jeta_cost,fuel_mogas_cost,notes'.split(",")]));
  }
  // If aircraft localStorage is empty, then set it to the default value
  if(window.localStorage.getItem("flights") == null){
    window.localStorage.setItem("flights", JSON.stringify(['airline,flight_number,route_code,callsign,route_leg,dpt_airport,arr_airport,alt_airport,days,dpt_time,arr_time,level,distance,flight_time,flight_type,load_factor,load_factor_variance,pilot_pay,route,notes,start_date,end_date,active,subfleets,fares,fields'.split(",")]));
  }
  // If aircraft localStorage is empty, then set it to the default value
  if(window.localStorage.getItem("pairs") == null){
    window.localStorage.setItem("pairs", JSON.stringify(['dep_icao,arr_icao,flights'.split(",")]));
  }

  // If the settings are empty, then run the setup
  await ipcRenderer.sendSync('getSettings');
  runSetup();
})

// * Function to run the setup
async function runSetup(){
  // Get the settings from the database
  let settings = JSON.parse(await ipcRenderer.sendSync('getSettings'));
  
  // Check if the fdpAPI settings are empty
  if(settings['fpdAPI'] == undefined){
    // If it is, then run the setup for fpdAPI
    alert("Please read 'https://bit.ly/fdbapi' before continuing. It includes guide for the next question.")
    let res = await ipcRenderer.sendSync("dialogCreate", "Setup", "Flight Plan Database API Key", "Documentation: https://github.com/virtualegret/flight-schedules/wiki/Getting-FlightPlanDatabase-API-Key", 'input')
    
    if(res == 501 || res == 404)return runSetup();
    let result = await ipcRenderer.sendSync("RESTreq", "https://api.flightplandatabase.com/me", "get", { 
      'Content-Type': 'application/json',
      'Authorization': 'Basic ' + res
    })
    if(typeof result == "number"){
      alert("An error(" + result + ") occured while trying to get the flight route. Please try again.");
      return runSetup();
    }
    
    // Save the settings to the database
    await ipcRenderer.sendSync('saveSettings', 'fpdAPI', res)
  }
}

// * Function to clear the settings
async function clearSettings(){
  await ipcRenderer.sendSync("clearSetup")
  runSetup()
}

// * Function to edit the settings
async function editSettings(){
  // Get the settings from the database
  let settings = JSON.parse(await ipcRenderer.sendSync('getSettings'));
  
  // Check if the fdpAPI settings are empty
  if(settings['fpdAPI'] == undefined){
    return alert("Please run the setup first!")
  }else{
    // If it is, then run the setup for fpdAPI
    let res = await ipcRenderer.sendSync("dialogCreate", "Setup", "Flight Plan Database API Key", settings['fpdAPI'], 'input')
    
    if(res == 404 || res == 501)return alert("Please enter a valid key.");
    let result = await ipcRenderer.sendSync("RESTreq", "https://api.flightplandatabase.com/me", "get", { 
      'Content-Type': 'application/json',
      'Authorization': 'Basic ' + res
    })
    if(typeof result == "number"){
      alert("An error(" + flightRoute + ") occured while trying to get the flight route. Please try again.");
      return
    }

    // Save the settings to the database
    await ipcRenderer.sendSync('saveSettings', 'fpdAPI', res)
  }
}

// * Alert dialog
// @param message - The message to be displayed in the dialog
function alert(message) {
    const options = {
      type: 'info',
      buttons: ['OK'],
      defaultId: 0,
      title: 'Alert',
      message: message == undefined ? 'Oops! This is not supposed to be sent! Please report this bug!' : message
    };
  
    ipcRenderer.sendSync('dialog', 'showMessageBox', options)
  }
  
  // * Confirm dialog
// @param message - The message to be displayed in the dialog
function confirm(message) {
  const options = {
    type: 'info',
    buttons: ['Yes', 'No'],
    defaultId: 1,
    title: 'Alert',
    message: message == undefined ? 'Oops! This is not supposed to be sent! Please report this bug!' : message
  };
  let msg = ipcRenderer.sendSync('dialog', 'showMessageBox', options)
  return msg.response
}

// * Function to go back to the menu
async function menu() {
  window.location.href = '../index.html'
}
  